packages/core/strapi/lib/services/entity-service/types/params/attributes.d.ts

```diff
@@ -56,6 +56,49 @@ export type ScalarValues = GetValue<
   | Attribute.UID<Common.UID.Schema | undefined>
 >;
 
+// TODO Add doc
+export type GetValues<TSchemaUID extends Common.UID.Schema> = OmitRelationWithoutTarget<
+  TSchemaUID,
+  {
+    [TKey in Attribute.GetOptionalKeys<TSchemaUID>]?: GetValue<Attribute.Get<TSchemaUID, TKey>>;
+  } & {
+    [TKey in Attribute.GetRequiredKeys<TSchemaUID>]-?: GetValue<Attribute.Get<TSchemaUID, TKey>>;
+  }
+>;
+
+export type OmitRelationWithoutTarget<TSchemaUID extends Common.UID.Schema, TValue> = Omit<
+  TValue,
+  Exclude<Attribute.GetKeysByType<TSchemaUID, 'relation'>, Attribute.GetKeysWithTarget<TSchemaUID>>
+>;
+
+export type ComponentValue<
```

Comment:
You could simplify by removing those helper types & do everything in the MatchFirst tests

```ts
      // Relation
      [
        Utils.Expression.Extends<TAttribute, Attribute.OfType<'relation'>>,
        // Reference relations that have target with IDs instead of values
        TAttribute extends Attribute.Relation<infer _TOrigin, infer TRelationKind, infer TTarget>
          ? Utils.Expression.If<
              Utils.Expression.IsNotNever<TTarget>,
              Attribute.RelationPluralityModifier<TRelationKind, `${number}` | number>
            >
          : never
      ],
      // DynamicZone
      [
        Utils.Expression.Extends<TAttribute, Attribute.OfType<'dynamiczone'>>,
        // Override the GetDynamicZoneValue to use the custom GetValues
        TAttribute extends Attribute.DynamicZone<infer TComponentsUIDs>
          ? // Extract tuple values to a component uid union type
            Utils.Array.Values<TComponentsUIDs> extends infer TComponentUID
            ? TComponentUID extends Common.UID.Component
              ? GetValues<TComponentUID> & { __component: TComponentUID }
              : never
            : never
          : never
      ],
      // Component
      [
        Utils.Expression.Extends<TAttribute, Attribute.OfType<'component'>>,
        // Override the GetComponentValue to use the custom GetValues
        TAttribute extends Attribute.Component<infer TComponentUID, infer TRepeatable>
          ? TComponentUID extends Common.UID.Component
            ? GetValues<TComponentUID> extends infer TValues
              ? Utils.Expression.If<TRepeatable, TValues[], TValues>
              : never
            : never
          : never
      ],
```